src/beast/include/beast/impl/basic_streambuf.ipp

```diff
@@ -478,74 +478,77 @@ auto
 basic_streambuf<Allocator>::prepare(size_type n) ->
     mutable_buffers_type
 {
-    iterator pos = out_;
-    if(pos != list_.end())
+    list_type reuse;
+    if(out_ != list_.end())
     {
-        auto const avail = pos->size() - out_pos_;
+        if(out_ != list_.iterator_to(list_.back()))
+        {
+            out_end_ = out_->size();
+            reuse.splice(reuse.end(), list_,
+                std::next(out_), list_.end());
+            debug_check();
+        }
+        auto const avail = out_->size() - out_pos_;
         if(n > avail)
         {
+            out_end_ = out_->size();
             n -= avail;
-            out_end_ = pos->size();
-            while(++pos != list_.end())
-            {
-                if(n < pos->size())
-                {
-                    out_end_ = n;
-                    n = 0;
-                    ++pos;
-                    break;
-                }
-                n -= pos->size();
-            }
         }
         else
         {
-            ++pos;
             out_end_ = out_pos_ + n;
             n = 0;
         }
+        debug_check();
     }
-
-    if(n > 0)
+    while(n > 0 && ! reuse.empty())
     {
-        assert(pos == list_.end());
-        for(;;)
+        auto& e = reuse.front();
+        reuse.erase(reuse.iterator_to(e));
+        list_.push_back(e);
+        if(n > e.size())
         {
-            auto const size = std::max(alloc_size_, n);
-            auto& e = *reinterpret_cast<element*>(
-                alloc_traits::allocate(this->member(),
-                    size + sizeof(element)));
-            alloc_traits::construct(this->member(), &e, size);
-            list_.push_back(e);
-            if(out_ == list_.end())
-            {
-                out_ = list_.iterator_to(e);
-                debug_check();
-            }
-            if(n <= size)
-            {
-                out_end_ = n;
-                debug_check();
-                break;
-            }
-            n -= size;
-            debug_check();
+            out_end_ = e.size();
+            n -= e.size();
+        }
+        else
+        {
+            out_end_ = n;
+            n = 0;
         }
+        debug_check();
     }
-    else
+    while(n > 0)
     {
-        while(pos != list_.end())
+        auto const size = std::max(alloc_size_, n);
+        auto& e = *reinterpret_cast<element*>(
+            alloc_traits::allocate(this->member(),
+                size + sizeof(element)));
```

Comment:
I think it would be useful to leave a comment noting why you are allocating `size + sizeof(element)`.  I'm not saying it's wrong.  But it's not obvious to even a non-casual reader why `sizeof(element)` is added to the allocation.